packages/core/src/html/UlNodeProcessor.ts

```diff
@@ -0,0 +1,189 @@
+  } else {
+    spanContent = processIconString(iconAttrs.icon);
+  }
+
+  let spanNode;
+  if (spanContent === null) {
+    const img = cheerio(`<img src='${iconAttrs.icon}' alt='Icon'>`)
+      .css({ width: iconAttrs.width, height: iconAttrs.height, display: 'inline-block' })
+      .addClass(iconAttrs.className || '');
+    img.append('\u200B');
+
+    spanContent = cheerio('<span></span>').append(img).css({
+      'padding-bottom': '0.3em',
+      'padding-top': '0.3em',
+    });
+    spanNode = cheerio(spanContent).css({ 'font-size': 'unset' });
+  } else {
+    spanNode = cheerio(spanContent).css({ 'font-size': 'unset' });
+    spanNode = spanNode.css({ 'font-size': iconAttrs.size }).addClass(iconAttrs.className || '');
+  }
+  // Add invisible character to avoid the element from being empty
+  spanNode.append('\u200B');
+  return spanNode.css({
+    'line-height': 'unset',
+    'margin-inline-end': '0.3em',
+    'flex-shrink': '0',
+    'align-self': 'flex-start',
+  });
+}
+
+function updateNodeStyle(node: NodeOrText) {
+  const nodeCheerio = cheerio(node);
+
+  nodeCheerio.css({
+    'list-style-type': 'none',
+    'padding-inline-start': '0px',
+  });
+}
+
+const getIconAttributes = (node: MbNode, defaultIcon?: IconAttributes): IconAttributes | null => {
+  // If node.attribs doesn't exist, return null immediately
+  if (!node.attribs || (defaultIcon?.icon === undefined && node.attribs.icon === undefined)) {
+    return null;
+  }
+
+  return {
+    icon: node.attribs.icon !== undefined ? node.attribs.icon : defaultIcon?.icon,
+    width: node.attribs['i-width'] !== undefined ? node.attribs['i-width'] : defaultIcon?.width,
+    height: node.attribs['i-height'] !== undefined ? node.attribs['i-height'] : defaultIcon?.height,
+    size: node.attribs['i-size'] !== undefined ? node.attribs['i-size'] : defaultIcon?.size,
+    className: node.attribs['i-class'] !== undefined ? node.attribs['i-class'] : defaultIcon?.className,
+  };
+};
+
+const deleteAttributes = (node: MbNode, attributes: string[]) => {
+  attributes.forEach((attr) => {
+    delete node.attribs[attr];
+  });
+};
+
+function updateLi(child: NodeOrText, iconAttributes: IconAttributes) {
+  const childNode = child as MbNode;
+  if (
+    iconAttributes.icon === undefined
+  ) return;
+  const curLiIcon = getIconAttributes(childNode, iconAttributes);
+
+  deleteAttributes(childNode, ICON_ATTRIBUTES);
+
+  // Create a new div and span
+  const div = cheerio('<div></div>');
+  const iconSpan = createIconSpan(curLiIcon!);
+
+  div.append(cheerio(child.children).remove());
+
+  // Append iconSpan and div to the child
+  cheerio(child).append(iconSpan);
+  cheerio(child).append(div);
+
+  cheerio(child).css({
+    display: 'flex',
+  });
+}
+
+function handleChild(child: MbNode, iconAttrValue: IconAttributeDetail) {
+  if (child.name === 'li') {
+    const liChild = child as MbNode;
+    if (iconAttrValue.isFirst) {
```

Comment:
Don't think this has been resolved yet -- its about the design choice of preventing custom icons from working at all if some icon in the list uses a default icon.

Reply:
How about this:
/**
 * Processes a `li` node. If icon attributes are specified, the function updates the node;
 * if not, the function leaves the node as is. This function prevents custom icons from 
 * being used when neither default icon attributes nor `node.attri.icon` are specified.
 *
 * @param {MbNode} node - The li node to process.
 * @param {IconAttributeDetail} iconAttrValue - Details of the icon attributes.
 */

Reply:
(small wording nit)
```diff
- This function prevents custom icons
+ This function also prevents
```

(provide context on the design choice, instead of the technicalities the code reader can figure out on their own as well)
```diff
- when neither default icon attributes nor node.attri.icon are specified.
+ when default list icons and node.attribs.icon are mixed.
```

Reply:
> (provide context on the design choice, instead of the technicalities the code reader can figure out on their own as well)

or both 😶 (but I think technically it is simple enough)

Reply:
> ```diff
> when default list icons and node.attribs.icon are mixed.
> ```

However, this isn't clear. When reading this, it's unclear what the default list and `node.attribs.icon` are. Furthermore, the term "mixing" is ambiguous, as we don't prohibit the combination of standard unordered lists with customized unordered lists. We only want to apply customization under specific conditions:
1. If an item has a specified icon, that icon will be used for it and for subsequent items at that level.
1. If any item at a level is customized, the first item at that level must also be customized. If not, the list will be invalidated.

Reply:
@SPWwj feel free to edit it, its a starting baseline.

E.g. `when the browser's default list icons and custom icons from node.attribs.icon are mixed for a given list level.`

Comments shouldn't redescribe what a code reader can easily gleam from reading the code.

Reply:
> If an item has a specified icon, that icon will be used for it and for subsequent items at that level.

I'd place this  at the start of the file as a separate comment if you'd like to include it, but I think it can be easily understood from the user guide.

> If any item at a level is customized, the first item at that level must also be customized. If not, the list will be invalidated.

Looks better but still somewhat of a technality -- What design choice / rationale does this behaviour serve?

E.g., addon: This is an intentional design choice in preventing the browser's default list icons and custom icons from node.attribs.icon are mixed from being mixed for a given list level, `<due to> <resign for design choice>` or `refer to this issue, ...`.

Reply:
> ~~but I think it can be easily understood from the user guide.~~


Given the file name is `UlNodeProcessor`, this could be useful, or -- we can rename the file after its feature, e.g. `CustomListIconProcessor / MediaListIconProcessor / ...`.

Reply:
I think this is better: 
This function ensure the first item of list at that level must also be customized. If not, the list will be invalidated and default bullets will be used.

Reply:
> This function ensure the first item of list at that level must also be customized. If not, the list will be invalidated and default bullets will be used.

Good additional detail on default bullets will be used, but still missing the _why_?

(Summarise the discussion you had with prof Damith earlier on this -- others reading the code do not have the context of this lengthy PR)

Reply:
I think this should explain the design choice for our implementation:
// This function ensure the first item of at that level must also be customized.
// If not, the list will be invalidated and default bullets will be used.
// This is to prevent unintentional mixing of standard and customized lists.

Reply:
```diff
+ // This function ensures the first item at that level must also be customized.
- // This function ensure the first item of at that level must also be customized.
// If not, the list will be invalidated and default bullets will be used.
// This is to prevent unintentional mixing of standard and customized lists.
+ // See https://github.com/MarkBind/markbind/pull/2316#discussion_r1255364486 for more details.
```

Add in a link for context, there is much more you both have pointed out earlier.

---

> CustomUlListProcessor.ts

The file name still dosen't tell much for someone without context: what sort of "custom" processing for `<ul>` is being done?

If you'd like to keep the filename we can also add a short comment at the start of the file as you've suggested.

e.g. 

```
/*
Processes lists to support custom list icons.
*/
```

Reply:
@SPWwj

Reply:
Sorry for missed that. I've changed the file name to `CustomListIconProcessor`. I believe this name suit the file better.